ocaml-lsp-server/src/code_actions/action_construct.ml

```diff
@@ -0,0 +1,54 @@
+open Import
+
+let action_kind = "construct"
+
+let code_action doc (params : CodeActionParams.t) =
+  let { Range.start = _; end_ } = params.range in
+  match Document.kind doc with
+  | Intf -> Fiber.return None
+  | Impl -> (
+    let src = Document.source doc in
+    let pos = Position.logical end_ in
+    (* we want this predicate to quickly eliminate prefixes that don't fit to be
+       a hole *)
+    let can_be_hole =
+      let prefix = Compl.prefix_of_position ~short_path:false src pos in
+      let hole_syntax = "_" in
+      String.equal prefix hole_syntax
+    in
+    if not can_be_hole then
+      Fiber.return None
+    else
+      let code_action =
+        let command : Command.t =
+          Command.create ~title:"Trigger Suggest"
+            ~command:"editor.action.triggerSuggest" ()
+        in
+        CodeAction.create ~title:"Construct an expression"
+          ~kind:(CodeActionKind.Other action_kind) ~command ()
+      in
+      Document.with_pipeline_exn doc @@ fun pipeline ->
```

Comment:
I'd prefer not to use the `@@` form. That makes it easier to slip in accidental work into the merlin thread as you've noticed in your first commit.